File: web/src/api/workflows.ts
import { apiClient, isTauri, tauriInvoke } from './config'
import type { Workflow } from '@/types/generated/Workflow'
import type { ExecutionHistoryPage } from '@/types/generated/ExecutionHistoryPage'
import { API_ENDPOINTS } from '@/constants'

export const listWorkflows = async (): Promise<Workflow[]> => {
  if (isTauri()) {
    return tauriInvoke<Workflow[]>('list_workflows')
  }
  const response = await apiClient.get<Workflow[]>(API_ENDPOINTS.WORKFLOW.LIST)
  return response.data
}

export const createWorkflow = async (workflow: Workflow): Promise<{ id: string }> => {
  if (isTauri()) {
    const result = await tauriInvoke<Workflow>('create_workflow', { workflow })
    return { id: result.id }
  }
  const response = await apiClient.post<{ id: string }>(API_ENDPOINTS.WORKFLOW.CREATE, workflow)
  return response.data
}

export const getWorkflow = async (id: string): Promise<Workflow> => {
  if (isTauri()) {
    return tauriInvoke<Workflow>('get_workflow', { id })
  }
  const response = await apiClient.get<Workflow>(API_ENDPOINTS.WORKFLOW.GET(id))
  return response.data
}

export const updateWorkflow = async (id: string, workflow: Workflow): Promise<void> => {
  if (isTauri()) {
    await tauriInvoke<Workflow>('update_workflow', { id, workflow })
    return
  }
  await apiClient.put<void>(API_ENDPOINTS.WORKFLOW.UPDATE(id), workflow)
}

export const deleteWorkflow = async (id: string): Promise<void> => {
  if (isTauri()) {
    return tauriInvoke<void>('delete_workflow', { id })
  }
  await apiClient.delete<void>(API_ENDPOINTS.WORKFLOW.DELETE(id))
}

export const submitWorkflow = async (
  id: string,
  initialVariables?: unknown,
): Promise<{ execution_id: string }> => {
  if (isTauri()) {
    const executionId = await tauriInvoke<string>('execute_workflow', {
      id,
      input: initialVariables ?? null,
    })
    return { execution_id: executionId }
  }
  const response = await apiClient.post<{ execution_id: string; workflow_id: string }>(
    API_ENDPOINTS.EXECUTION.SUBMIT(id),
    { initial_variables: initialVariables },
  )
  return { execution_id: response.data.execution_id }
}

export const listWorkflowExecutions = async (
  id: string,
  page = 1,
  pageSize = 20,
): Promise<ExecutionHistoryPage> => {
  if (isTauri()) {
    return tauriInvoke<ExecutionHistoryPage>('get_workflow_executions', {
      workflow_id: id,
      page,
      page_size: pageSize,
    })
  }
  const response = await apiClient.get<ExecutionHistoryPage>(API_ENDPOINTS.EXECUTION.HISTORY(id), {
    params: { page, page_size: pageSize },
  })
  return response.data
}

File: crates/restflow-tauri/src/commands/workflows.rs
//! Workflow-related Tauri commands

use crate::state::AppState;
use restflow_workflow::models::ExecutionHistoryPage;
use restflow_workflow::Workflow;
use serde_json::Value;
use tauri::State;

/// List all workflows
#[tauri::command]
pub async fn list_workflows(state: State<'_, AppState>) -> Result<Vec<Workflow>, String> {
    state
        .core
        .storage
        .workflows
        .list_workflows()
        .map_err(|e| e.to_string())
}

/// Get a workflow by ID
#[tauri::command]
pub async fn get_workflow(state: State<'_, AppState>, id: String) -> Result<Workflow, String> {
    state
        .core
        .storage
        .workflows
        .get_workflow(&id)
        .map_err(|e| e.to_string())
}

/// Create a new workflow
#[tauri::command]
pub async fn create_workflow(
    state: State<'_, AppState>,
    workflow: Workflow,
) -> Result<Workflow, String> {
    state
        .core
        .storage
        .workflows
        .create_workflow(&workflow)
        .map_err(|e| e.to_string())?;

    Ok(workflow)
}

/// Update an existing workflow
#[tauri::command]
pub async fn update_workflow(
    state: State<'_, AppState>,
    id: String,
    workflow: Workflow,
) -> Result<Workflow, String> {
    // Ensure the ID matches
    let mut updated_workflow = workflow;
    updated_workflow.id = id.clone();

    state
        .core
        .storage
        .workflows
        .update_workflow(&id, &updated_workflow)
        .map_err(|e| e.to_string())?;

    Ok(updated_workflow)
}

/// Delete a workflow by ID
#[tauri::command]
pub async fn delete_workflow(state: State<'_, AppState>, id: String) -> Result<(), String> {
    state
        .core
        .storage
        .workflows
        .delete_workflow(&id)
        .map_err(|e| e.to_string())
}

/// Execute a workflow
#[tauri::command]
pub async fn execute_workflow(
    state: State<'_, AppState>,
    id: String,
    input: Option<Value>,
) -> Result<String, String> {
    let input_value = input.unwrap_or(Value::Null);

    state
        .core
        .executor
        .submit(id, input_value)
        .await
        .map_err(|e| e.to_string())
}

/// Get workflow execution history (paginated)
#[tauri::command]
pub async fn get_workflow_executions(
    state: State<'_, AppState>,
    workflow_id: String,
    page: Option<usize>,
    page_size: Option<usize>,
) -> Result<ExecutionHistoryPage, String> {
    let page = page.unwrap_or(1);
    let page_size = page_size.unwrap_or(50);

    state
        .core
        .storage
        .execution_history
        .list_paginated(&workflow_id, page, page_size)
        .map_err(|e| e.to_string())
}
